### PR TITLE
Fix conditional joins with empty left table

### LINKED ISSUE #9044: [BUG] Conditional join with empty left table throws exception
**Describe the bug**
Performing a conditional join when the left table has zero rows throws the following exception:
```
cuDF failure at: ../include/cudf/detail/utilities/cuda.cuh:65: num_blocks must be > 0
```

**Steps/Code to reproduce bug**
Apply the following patch and run JOIN_TEST:
```
diff --git a/cpp/tests/join/conditional_join_tests.cu b/cpp/tests/join/conditional_join_tests.cu
index e16e1ec7de..ad70fe9a46 100644
--- a/cpp/tests/join/conditional_join_tests.cu
+++ b/cpp/tests/join/conditional_join_tests.cu
@@ -563,6 +563,16 @@ TYPED_TEST(ConditionalFullJoinTest, TestOneColumnNoneEqual)
               {JoinNoneValue, 2}});
 };
 
+TYPED_TEST(ConditionalFullJoinTest, TestOneColumnLeftEmpty)
+{
+  this->test({{}},
+             {{3, 4, 5}},
+             left_zero_eq_right_zero,
+             {{JoinNoneValue, 0},
+              {JoinNoneValue, 1},
+              {JoinNoneValue, 2}});
+};
+
 TYPED_TEST(ConditionalFullJoinTest, TestTwoColumnThreeRowSomeEqual)
 {
   this->test({{0, 1, 2}, {10, 20, 30}},
```

Note that this fails for other join types as well such as left and inner joins.

**Expected behavior**
Joins should not crash when given one or more empty tables as input.

### PR DESCRIPTION
**Pull request by vyasr**
Resolves #9044 by ensuring that when the left table is empty the functions short-circuit accordingly. Where necessary (i.e. for full joins) the complementary pairs of indices are returned.

### CODE REVIEW

**Comment by vyasr** [comment]
@jlowe this should unblock nightly tests.

File: cpp/src/join/conditional_join.cu [hunk 1]
```diff
@@ -62,6 +61,21 @@ conditional_join(table_view const& left,
...
+      case join_kind::LEFT_ANTI_JOIN:
+      case join_kind::INNER_JOIN:
+      case join_kind::LEFT_SEMI_JOIN:
+        return std::make_pair(std::make_unique<rmm::device_uvector<size_type>>(0, stream, mr),
+                              std::make_unique<rmm::device_uvector<size_type>>(0, stream, mr));
+      // Full joins need to return the trivial complement.
+      case join_kind::FULL_JOIN: {
+        auto ret_flipped = get_trivial_left_join_indices(right, stream);
+        return std::make_pair(std::move(ret_flipped.second), std::move(ret_flipped.first));
+      }
```
Comment (ttnghia) [inline cpp/src/join/conditional_join.cu:77]: This `switch` is lacking of the `default` case. We don't need it but the compiler may issue a warning which can be treated as error. So just add a line `default: ;`.

File: cpp/src/join/conditional_join.cu [hunk 2]
```diff
@@ -118,8 +132,14 @@ conditional_join(table_view const& left,
...
   if (join_size == 0) {
-    return std::make_pair(std::make_unique<rmm::device_uvector<size_type>>(0, stream, mr),
-                          std::make_unique<rmm::device_uvector<size_type>>(0, stream, mr));
+    auto join_indices{
+      std::make_pair(std::make_unique<rmm::device_uvector<size_type>>(0, stream, mr),
+                     std::make_unique<rmm::device_uvector<size_type>>(0, stream, mr))};
+    if (join_type == join_kind::FULL_JOIN) {
+      auto complement_indices = detail::get_left_join_indices_complement(
+        join_indices.second, left.num_rows(), right.num_rows(), stream, mr);
+      join_indices = detail::concatenate_vector_pairs(join_indices, complement_indices, stream);
```
Comment (ttnghia) [inline cpp/src/join/conditional_join.cu:141]: Wait, I see that `join_indices` is a local variable and it is not used anywhere?

File: cpp/src/join/conditional_join.cu [hunk 3]
```diff
@@ -177,21 +197,30 @@ std::size_t compute_conditional_join_output_size(table_view const& left,
...
+  } else if (left.num_rows() == 0) {
+    switch (join_type) {
+      // Left, left anti, left semi, and inner joins all return empty sets.
+      case join_kind::LEFT_JOIN:
+      case join_kind::LEFT_ANTI_JOIN:
+      case join_kind::INNER_JOIN:
+      case join_kind::LEFT_SEMI_JOIN: return 0;
+      // Full joins need to return the trivial complement.
+      case join_kind::FULL_JOIN: return right.num_rows();
+    }
```
Comment (ttnghia) [inline cpp/src/join/conditional_join.cu:223]: Similar to above for these 2 `switch`.

File: cpp/src/join/conditional_join.cu (hunk 2 above)
Comment (ttnghia) [reply cpp/src/join/conditional_join.cu:141]: Previously there is a `return` statement here but it has been deleted. Is this intentional?

Comment (vyasr) [reply cpp/src/join/conditional_join.cu:141]: Good catch, this is a mistake on my part. Fortunately (or maybe unfortunately?) the code still generates the correct result without it, it just does more work first. Fixed.

Comment (ttnghia) [reply cpp/src/join/conditional_join.cu]: I realized that `nvcc` does not catch all warnings as g++, which eventually be treated as errors so we can detect the  issue like this immediately. A lot of other ignored warnings were just recently fixed.

File: cpp/src/join/conditional_join.cu [hunk 4]
```diff
@@ -118,8 +134,14 @@ conditional_join(table_view const& left,
...
-    return std::make_pair(std::make_unique<rmm::device_uvector<size_type>>(0, stream, mr),
-                          std::make_unique<rmm::device_uvector<size_type>>(0, stream, mr));
+    auto join_indices{
+      std::make_pair(std::make_unique<rmm::device_uvector<size_type>>(0, stream, mr),
+                     std::make_unique<rmm::device_uvector<size_type>>(0, stream, mr))};
+    if (join_type == join_kind::FULL_JOIN) {
+      auto complement_indices = detail::get_left_join_indices_complement(
+        join_indices.second, left.num_rows(), right.num_rows(), stream, mr);
+      return detail::concatenate_vector_pairs(join_indices, complement_indices, stream);
+    }
```
Comment (ttnghia) [inline cpp/src/join/conditional_join.cu:144]: Here I only see "If the output size is zero AND join_type == join_kind::FULL_JOIN, we can return immediately" .

Comment (vyasr) [reply cpp/src/join/conditional_join.cu]: Wow thank you for bothering me more on this, clearly my last fix was hasty. I've added a much more descriptive comment of what the behavior _should_ be and updated the code to match.

**Review by ttnghia — APPROVED** [review]

**Comment by vyasr** [comment]
rerun tests

**Comment by vyasr** [comment]
rerun tests

**Comment by vyasr** [comment]
@gpucibot merge

**Pull request merged**